test/fixtures/es-module-loaders/loader-with-custom-condition.mjs

```diff
@@ -0,0 +1,11 @@
+import {ok} from 'assert';
+
+export async function resolve(specifier, context, defaultResolve) {
+  ok(Array.isArray(context.conditions), 'loader receives conditions array');
+  ok(context.conditions.includes('node'), 'provided conditions include "node"');
+  ok(!context.conditions.includes('default'), 'default is implicit');
```

Comment:
So is `context.conditions` always an object with keys `node` and `import` (per `const DEFAULT_CONDITIONS = ObjectFreeze(['node', 'import'])` above)? That seems . . . weird. Why not `import` and `require` if you’re going to do that, or `import` and `require` and `node` and `default` if we’re going to list all the conditions that Node itself supports?

Or just only list the conditions actually defined in the `package.json`? If conditional exports syntax isn’t used, I guess you would just do the one `default` condition (as not using the conditional syntax is the same as using it with the `default` condition, I think). And if conditional exports syntax _is_ used, include only the conditions that the user has defined.

Reply:
> Why not import and require if you’re going to do that, or import and require and node and default if we’re going to list all the conditions that Node itself supports?

This code only runs (right now) for imports. So it shouldn't include `require` because then it would match `require` conditions for import. It lists all conditions that apply to this resolution request and can potentially be changed by the loader hook. I made the choice of not allowing the hook to mess with `default`.

If at some future point `require` resolution should pass through loader hooks, then those resolutions would have `conditions` set to `["node", "require"]` by default.

I'm not sure I follow the rest of your comment. The conditions are the ones that are active for this resolution. Nothing in any `package.json` will affect this. This list is potentially used to match the conditions in `package.json`. But there's no way to know which conditions will actually appear in any `package.json` file (or if any `package.json` file will be loaded in the first place) when the `resolve` hook is called.

Reply:
I'm confused. When you say that we're exposing the exports conditions, I assumed that what you meant was that for a particular `"exports"` left-hand-side, the entire right-hand-side object would be provided to the loader. Not just the particular condition that matched. Wouldn't we want all the information, so that the loader could choose on its own to pick a different condition? E.g. if someone made a `"typescript"` condition say, we'd want the loader to know that that exists and read it and choose to load that instead of the matched `"import"` or `"node"` or `"default"` condition.

Reply:
This PR is to allow loaders to pass around a `"typescript"` condition; it doesn't affect resolution, but does allow custom loaders to alter the conditions that are being resolved thereby altering resolution by adding/removing/etc. the set of conditions.

I don't understand how the right-hand-side of `"exports"` comes into play here as the PR is only about the left hand side and it is up to loaders to resolve the right hand side.

Reply:
When I say left-hand-side, I'm referring to the path, e.g. the `"./feature"` in:

```js
  "exports": {
    "./feature": {
      "import": "./feature-default.js",
      "browser": "./feature-browser.js"
    }
```

Reply:
Ah, this PR is about the "import"/"browser" LHS. We were talking about different things!

Reply:
> E.g. if someone made a "typescript" condition say, we'd want the loader to know that that exists and read it and choose to load that instead of the matched "import" or "node" or "default" condition.

I think you're thinking about it from the opposite direction. The `resolve` comes before any `package.json` file has been loaded. There is no `exports` yet. There's only "what conditions are active for this resolution", not yet "what mappings are possible". So hook doesn't have to know the latter. If it wants to allow a `browser` or `typescript` mapping, it just needs to it to the conditions when it delegates to the parent hook.

Example realistic-ish hooks:

```js
export function resolve(specifier, context, next) {
  return next(specifier, {
    ...context,
    {
      conditions: [
        process.env.NODE_ENV === 'production' ? 'production' : 'development',
        ...context.conditions,
      ],
    }
  });
}
```

1. Custom resolve hook is called, it's from an import so we pass `["node", "import"]`.
2. The hook wants to add either `production` or `development` to the recognized conditions, so it calls the parent with those.
3. The built-in resolve logic searches for a `package.json` and will match using all the conditions passed by the hook.

I don't see how the hook could know about contents of the `package.json` and what kind of use case you are thinking about when you're talking about those keys.

A hook that uses the conditions itself, not just changes which are active:

```js
const CACHED_IMPORT_MAPS = loadImportMaps();

export function resolve(specifier, context, next) {
  const conditionsKey = context.conditions.slice(0).sort().join('\0');
  if (CACHED_IMPORT_MAPS.has(conditionsKey) {
    return resolveFromImportMap(
      specifier,
      context,
      CACHED_IMPORT_MAPS.get(conditionsKey)
    );
  }
  return next(specifier, context, next);
}
```

In neither of those cases it seems valuable to pass conditions extracted from a `package.json` to the hook.

Reply:
I'm not really understanding the purpose of this PR then. It would seem that the exposed conditions would always be `["node", "import"]`, since we currently don't have composition of loaders. In other words,

> The built-in resolve logic searches for a `package.json` and will match using all the conditions passed by the hook.

But then the Node default `resolve` would always ignore a `production` or `development` condition, so what purpose would have been served by adding such a custom condition? It would seem to be useful only if there was some other loader next in the chain that then would do something with this custom-added condition. Right? But there can't be one, at least not yet.

Reply:
Node as of this PR would process all of the Set rather than just "import" and "node", thats what https://github.com/nodejs/node/pull/31303/files#diff-4a5950be44f56dfca2e7fa3c4e1fc0e4R1058 is changing